frame/support/src/storage/types/nmap.rs

```diff
@@ -0,0 +1,990 @@
+	///
+	/// If you add or remove values to the map while doing this, you'll get undefined results.
+	pub fn iter() -> crate::storage::PrefixIterator<(Key::Key, Value)> {
+		<Self as crate::storage::IterableStorageNMap<Key, Value>>::iter()
+	}
+
+	/// Remove all elements from the map and iterate through them in no particular order.
+	///
+	/// If you add elements to the map while doing this, you'll get undefined results.
+	pub fn drain() -> crate::storage::PrefixIterator<(Key::Key, Value)> {
+		<Self as crate::storage::IterableStorageNMap<Key, Value>>::drain()
+	}
+
+	/// Translate the values of all elements by a function `f`, in the map in no particular order.
+	///
+	/// By returning `None` from `f` for an element, you'll remove it from the map.
+	///
+	/// NOTE: If a value fail to decode because storage is corrupted then it is skipped.
+	pub fn translate<O: Decode, F: FnMut(Key::Key, O) -> Option<Value>>(f: F) {
+		<Self as crate::storage::IterableStorageNMap<Key, Value>>::translate(f)
+	}
+}
+
+/// Part of storage metadata for a storage n map.
+///
+/// NOTE: Generic hashers is supported.
+pub trait StorageNMapMetadata {
+	const MODIFIER: StorageEntryModifier;
+	const NAME: &'static str;
+	const DEFAULT: DefaultByteGetter;
+}
+
+impl<Prefix, Key, Value, QueryKind, OnEmpty> StorageNMapMetadata
+	for StorageNMap<Prefix, Key, Value, QueryKind, OnEmpty>
+where
+	Prefix: StorageInstance,
+	Key: super::key::KeyGenerator,
+	Value: FullCodec,
+	QueryKind: QueryKindTrait<Value, OnEmpty>,
+	OnEmpty: crate::traits::Get<QueryKind::Query> + 'static,
+{
+	const MODIFIER: StorageEntryModifier = QueryKind::METADATA;
+	const NAME: &'static str = Prefix::STORAGE_PREFIX;
+	const DEFAULT: DefaultByteGetter = DefaultByteGetter(
+		&OnEmptyGetter::<QueryKind::Query, OnEmpty>(core::marker::PhantomData),
+	);
+}
+
+#[cfg(test)]
+mod test {
+	use super::*;
+	use crate::hash::*;
+	use crate::storage::types::{Key, ValueQuery};
+	use frame_metadata::StorageEntryModifier;
+	use sp_io::{hashing::twox_128, TestExternalities};
+
+	struct Prefix;
+	impl StorageInstance for Prefix {
+		fn pallet_prefix() -> &'static str {
+			"test"
+		}
+		const STORAGE_PREFIX: &'static str = "foo";
+	}
+
+	struct ADefault;
+	impl crate::traits::Get<u32> for ADefault {
+		fn get() -> u32 {
+			98
+		}
+	}
+
+	#[test]
+	fn test_1_key() {
+		type A = StorageNMap<Prefix, Key<Blake2_128Concat, u16>, u32, OptionQuery>;
+		type AValueQueryWithAnOnEmpty =
+			StorageNMap<Prefix, Key<Blake2_128Concat, u16>, u32, ValueQuery, ADefault>;
+		type B = StorageNMap<Prefix, Key<Blake2_256, u16>, u32, ValueQuery>;
+		type C = StorageNMap<Prefix, Key<Blake2_128Concat, u16>, u8, ValueQuery>;
+		type WithLen = StorageNMap<Prefix, Key<Blake2_128Concat, u16>, Vec<u32>>;
+
+		TestExternalities::default().execute_with(|| {
+			let mut k: Vec<u8> = vec![];
+			k.extend(&twox_128(b"test"));
+			k.extend(&twox_128(b"foo"));
+			k.extend(&3u16.blake2_128_concat());
+			assert_eq!(A::hashed_key_for((3,)).to_vec(), k);
```

Comment:
we could also make use of EncodeLike to ensure it is working as expected
```suggestion
			assert_eq!(A::hashed_key_for((&3,)).to_vec(), k);
```